src/Init/Data/Ord.lean

```diff
@@ -530,13 +532,225 @@ instance : Ord USize where
+
+protected theorem isLE_compare_nil_left {α} [Ord α] {xs : List α} :
+    (compare [] xs).isLE :=
+  List.isLE_compareLex_nil_left
+
+protected theorem isLE_compareLex_nil_right {α} {cmp} {xs : List α} :
+    (List.compareLex (cmp := cmp) xs []).isLE ↔ xs = [] := by
+  cases xs <;> simp [List.compareLex_nil_nil, List.compareLex_cons_nil]
+
+@[simp]
+protected theorem isLE_compare_nil_right {α} [Ord α] {xs : List α} :
+    (compare xs []).isLE ↔ xs = [] :=
+  List.isLE_compareLex_nil_right
+
+protected theorem isGE_compareLex_nil_left {α} {cmp} {xs : List α} :
+    (List.compareLex (cmp := cmp) [] xs).isGE ↔ xs = [] := by
+  cases xs <;> simp [List.compareLex_nil_nil, List.compareLex_nil_cons]
+
+@[simp]
+protected theorem isGE_compare_nil_left {α} [Ord α] {xs : List α} :
+    (compare [] xs).isGE ↔ xs = [] :=
+  List.isGE_compareLex_nil_left
+
+protected theorem isGE_compareLex_nil_right {α} {cmp} {xs : List α} :
+    (List.compareLex (cmp := cmp) xs []).isGE := by
+  cases xs <;> simp [List.compareLex_nil_nil, List.compareLex_cons_nil]
+
+protected theorem isGE_compare_nil_right {α} [Ord α] {xs : List α} :
+    (compare xs []).isGE :=
+  List.isGE_compareLex_nil_right
+
+protected theorem compareLex_nil_left_eq_eq {α} {cmp} {xs : List α} :
+    List.compareLex cmp [] xs = .eq ↔ xs = [] := by
+  cases xs <;> simp [List.compareLex_nil_nil, List.compareLex_nil_cons]
+
+@[simp]
+protected theorem compare_nil_left_eq_eq {α} [Ord α] {xs : List α} :
+    compare [] xs = .eq ↔ xs = [] :=
+  List.compareLex_nil_left_eq_eq
+
+protected theorem compareLex_nil_right_eq_eq {α} {cmp} {xs : List α} :
+    xs.compareLex cmp [] = .eq ↔ xs = [] := by
+  cases xs <;> simp [List.compareLex_nil_nil, List.compareLex_cons_nil]
+
+@[simp]
+protected theorem compare_nil_right_eq_eq {α} [Ord α] {xs : List α} :
+    compare xs [] = .eq ↔ xs = [] :=
+  List.compareLex_nil_right_eq_eq
+
+end List
+
+namespace Array
+
+@[specialize]
+protected def compareLex {α} (cmp : α → α → Ordering) (a₁ a₂ : Array α) : Ordering :=
+  go 0
+where go i :=
+  if h₁ : a₁.size <= i then
+    if a₂.size <= i then .eq else .lt
+  else
+    if h₂ : a₂.size <= i then
+      .gt
+    else match cmp a₁[i] a₂[i] with
+      | .lt => .lt
+      | .eq => go (i + 1)
+      | .gt => .gt
+termination_by a₁.size - i
+
+instance {α} [Ord α] : Ord (Array α) where
+  compare := Array.compareLex compare
+
+protected theorem compare_eq_compareLex {α} [Ord α] :
+    compare (α := Array α) = Array.compareLex compare := rfl
+
+private theorem compareLexGo_succ {α} {cmp} {x₁ x₂} {a₁ a₂ : List α} {i} :
+    compareLex.go cmp (x₁ :: a₁).toArray (x₂ :: a₂).toArray (i + 1) =
+      compareLex.go cmp a₁.toArray a₂.toArray i := by
+  induction i using Array.compareLex.go.induct cmp a₁.toArray a₂.toArray
+  all_goals try
+    conv => congr <;> rw [compareLex.go]
+    simp
+    repeat' split <;> (try simp_all; done)
+
+private theorem compareLex_toArray_eq_listCompareLex {α} {cmp} {l₁ l₂ : List α} :
```

Comment:
Why is this `private`? It seems like a useful variant of the lemma below to me. And there should also be a `compare` version.

Reply:
Agreed, but I decided to rename this to `_root_.List.compareLex_eq_compareLex_toArray`, which seems more nice and more consistent to me.